%Version 3.1 December 2024
% See section 11 of the User Manual for version history
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%\documentclass[referee,sn-basic]{sn-jnl}% referee option is meant for double line spacing

%%=======================================================%%
%% to print line numbers in the margin use lineno option %%
%%=======================================================%%

%%\documentclass[lineno,pdflatex,sn-basic]{sn-jnl}% Basic Springer Nature Reference Style/Chemistry Reference Style

%%=========================================================================================%%
%% the documentclass is set to pdflatex as default. You can delete it if not appropriate.  %%
%%=========================================================================================%%

%%\documentclass[sn-basic]{sn-jnl}% Basic Springer Nature Reference Style/Chemistry Reference Style

%%Note: the following reference styles support Namedate and Numbered referencing. By default the style follows the most common style. To switch between the options you can add or remove Numbered in the optional parenthesis. 
%%The option is available for: sn-basic.bst, sn-chicago.bst%  
 
%%\documentclass[pdflatex,sn-nature]{sn-jnl}% Style for submissions to Nature Portfolio journals
%%\documentclass[pdflatex,sn-basic]{sn-jnl}% Basic Springer Nature Reference Style/Chemistry Reference Style
\documentclass[pdflatex,sn-mathphys-num]{sn-jnl}% Math and Physical Sciences Numbered Reference Style
%%\documentclass[pdflatex,sn-mathphys-ay]{sn-jnl}% Math and Physical Sciences Author Year Reference Style
%%\documentclass[pdflatex,sn-aps]{sn-jnl}% American Physical Society (APS) Reference Style
%%\documentclass[pdflatex,sn-vancouver-num]{sn-jnl}% Vancouver Numbered Reference Style
%%\documentclass[pdflatex,sn-vancouver-ay]{sn-jnl}% Vancouver Author Year Reference Style
%%\documentclass[pdflatex,sn-apa]{sn-jnl}% APA Reference Style
%%\documentclass[pdflatex,sn-chicago]{sn-jnl}% Chicago-based Humanities Reference Style

%%%% Standard Packages
%%<additional latex packages if required can be included here>

\usepackage{graphicx}%
\usepackage{multirow}%
\usepackage{amsmath,amssymb,amsfonts}%
\usepackage{amsthm}%
% Coleman
% Fix "Font shape `U/rsfs/m/n' in size <8.43146> not available"
% \usepackage{mathrsfs}%
\usepackage[title]{appendix}%
\usepackage{xcolor}%
\usepackage{textcomp}%
\usepackage{manyfoot}%
\usepackage{booktabs}%
\usepackage{algorithm}%
\usepackage{algorithmicx}%
\usepackage{algpseudocode}%
\usepackage{listings}%

% Coleman
\usepackage{makecell}
%%%%

%%%%%=============================================================================%%%%
%%%%  Remarks: This template is provided to aid authors with the preparation
%%%%  of original research articles intended for submission to journals published 
%%%%  by Springer Nature. The guidance has been prepared in partnership with 
%%%%  production teams to conform to Springer Nature technical requirements. 
%%%%  Editorial and presentation requirements differ among journal portfolios and 
%%%%  research disciplines. You may find sections in this template are irrelevant 
%%%%  to your work and are empowered to omit any such section if allowed by the 
%%%%  journal you intend to submit to. The submission guidelines and policies 
%%%%  of the journal take precedence. A detailed User Manual is available in the 
%%%%  template package for technical guidance.
%%%%%=============================================================================%%%%

%% as per the requirement new theorem styles can be included as shown below
\theoremstyle{thmstyleone}%
\newtheorem{theorem}{Theorem}%  meant for continuous numbers
%%\newtheorem{theorem}{Theorem}[section]% meant for sectionwise numbers
%% optional argument [theorem] produces theorem numbering sequence instead of independent numbers for Proposition
\newtheorem{proposition}[theorem]{Proposition}% 
%%\newtheorem{proposition}{Proposition}% to get separate numbers for theorem and proposition etc.

\theoremstyle{thmstyletwo}%
\newtheorem{example}{Example}%
\newtheorem{remark}{Remark}%

\theoremstyle{thmstylethree}%
\newtheorem{definition}{Definition}%

\raggedbottom
%%\unnumbered% uncomment this for unnumbered level heads

% Coleman
% Multi-file
% https://www.overleaf.com/learn/latex/Multi-file_LaTeX_projects
\usepackage{blindtext}
\usepackage{subfiles} % Best loaded last in the preamble
\begin{document}
% Coleman
% \title[short title]{full title}
\title[MTSCCleav]{MTSCCleav: a Multivariate Time Series Classification (MTSC)-based method for predicting human Dicer cleavage sites }

%%=============================================================%%
%% GivenName	-> \fnm{Joergen W.}
%% Particle	-> \spfx{van der} -> surname prefix
%% FamilyName	-> \sur{Ploeg}
%% Suffix	-> \sfx{IV}
%% \author*[1,2]{\fnm{Joergen W.} \spfx{van der} \sur{Ploeg} 
%%  \sfx{IV}}\email{iauthor@gmail.com}
%%=============================================================%%

\author*[1,2]{\fnm{First} \sur{Author}}\email{iauthor@gmail.com}

\author[2,3]{\fnm{Second} \sur{Author}}\email{iiauthor@gmail.com}
\equalcont{These authors contributed equally to this work.}

\author[1,2]{\fnm{Third} \sur{Author}}\email{iiiauthor@gmail.com}
\equalcont{These authors contributed equally to this work.}

\affil*[1]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \city{City}, \postcode{100190}, \state{State}, \country{Country}}}

\affil[2]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \city{City}, \postcode{10587}, \state{State}, \country{Country}}}

\affil[3]{\orgdiv{Department}, \orgname{Organization}, \orgaddress{\street{Street}, \city{City}, \postcode{610101}, \state{State}, \country{Country}}}

% \author*[1]{\fnm{Coleman} \sur{Yu}}\email{cyu@kuicr.kyoto-u.ac.jp}

% \author[2]{\fnm{Raymond Chi-Wing} \sur{Wong}}\email{raywong@cse.ust.hk}

% \author[1]{\fnm{Tomoya} \sur{Mori}}\email{tmori@kuicr.kyoto-u.ac.jp}

% \author[1]{\fnm{Tatsuya} \sur{Akutsu}}\email{takutsu@kuicr.kyoto-u.ac.jp}


% \affil*[1]{\orgdiv{Bioinformatics Center, Institute for Chemical Research}, \orgname{Kyoto University}, \orgaddress{Uji, Kyoto 611-0011, Japan}}

% \affil[2]{\orgdiv{Department of Computer
% Science and Engineering}, \orgname{The Hong Kong University of Science and Technology}, \orgaddress{Clear Water Bay, Kowloon, Hong Kong}}

%%==================================%%
%% Sample for unstructured abstract %%
%%==================================%%

%%================================%%
%% Sample for structured abstract %%
%%================================%%

\abstract{\textbf{Background:} MicroRNAs (miRNAs) are small non-coding RNAs (ncRNAs) that regulate gene expression at the post-transcriptional level and hence play essential roles in diverse biological processes such as development and differentiation.
The biogenesis of miRNAs requires Dicer, which is an enzyme, to cleave at specific sites on the precursor miRNAs (pre-miRNAs).
These sites are called cleavage sites.
Several machine learning approaches, such as ReCGBM and DiCleave, have been proposed to predict human dicer cleavage sites.
Given an input sequence, these classifiers predict whether it contains a cleavage site.
Despite the advances made, existing studies have several limitations.
They have ignored the probabilities of the base pairs in the secondary structure predicted by RNA secondary structure prediction tools in the classification.
Besides, they rely heavily on complicated feature engineering or opaque deep neural models. It results in a lack of generalizability and interoperability. Also, they have a long running time.
There is a need for alternative modeling paradigms that are simple, fast, and provide comparable accuracy while offering better model transparency.

\textbf{Results:} We propose a novel approach to predict human Dicer cleavage sites by reframing the task as a multivariate time series classification problem. 
To reframe, we have proposed different transformation schemes to convert the RNA sequence and the information about its secondary structure into time series.
Hence, the data can be represented in the form of a multivariate time series.
We also proposed a novel transformation scheme that involves the probabilities of the base pairs in the predicted secondary structure, which have been long ignored in the literature.
Computational experiments show that our proposed scheme can achieve comparable results using a simpler, intuitive model and less computation time.
Besides, we test our models with perturbation-based experiments. We found that the regions or subsequences close to the cleavage sites and hence to the center of pre-miRNAs are essential to the predictions of the human dicer cleavage sites.
Of note, the transformation of RNA data to time series allows the use of many state-of-the-art algorithms in the time series community and can relate some novel problem definitions in time series, such as motifs, discords, and chains, to the computational study of RNA data that paves novel ways using the well-established tools in the time series community.

\textbf{Conclusion:} Our proposed scheme allows us to study this problem in a new way. By transforming the RNA sequence and its secondary structure information into time series and using simple state-of-the-art time series classifiers, we obtain comparable or even superior performance in a simpler, faster way.
We introduce a novel RNA transformation method that leverages the base pair probabilities.
We also analyze the importance of the subsequences of the multivariate time series to the classification task, which hints that the regions that are close to the center are essential for this problem.
Code is available at: \href{https://github.com/cyuab/time-series-classification-cleavage}{https://github.com/cyuab/time-series-classification-cleavage}.
}

\keywords{miRNA, Dicer Cleavage Site, Genomic signal processing (GSP), (Multivariate) time Series Classification (MTSC, TSC)}

%%\pacs[JEL Classification]{D8, H51}

%%\pacs[MSC Classification]{35A01, 65L10, 65L12, 65L20, 65L70}

\maketitle

% Testing for bibliography~\cite{DiCleaveDeepLearning2024}.
\subfile{sections/background} % 1500 words
% \subfile{sections/methods} % 1500 words
% \subfile{sections/results} % 500 words
% \subfile{sections/discussion} % 500 words
% \subfile{sections/conclusions} % 500 words


%%===========================================================================================%%
%% If you are submitting to one of the Nature Portfolio journals, using the eJP submission   %%
%% system, please include the references within the manuscript file itself. You may do this  %%
%% by copying the reference list from your .bbl file, paste it into the main manuscript .tex %%
%% file, and delete the associated \verb+\bibliography+ commands.                            %%
%%===========================================================================================%%

\bibliography{main}% common bib file
%% if required, the content of .bbl file can be included here once bbl is generated
%%\input sn-article.bbl

\end{document}
